keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,662 @@
+For example, Windows Server 2019 container images can run inside a Utility VM (UVM) on a Windows Server 2022 host and because of this the image pulls should be allowed. For the full matrix of accepted behavior, check the link here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility
+
+` Note: ` UVM is short for utility VM. It is a light weight VM that boots a special version of Windows that is purpose built to do just one thing- run containers. UVMs provide strong isolation boundaries and do not share kernel with the host OS. They are transparent to the user and do not require any management from the user.
+
+This problem can be fixed if container runtimes, CRI, and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and the pull implementation on container runtimes can be extended to pull the appropriate manifest from the image index.
+An example implementation in the containerd container runtime on how the pull implementation can be extended to support this is as follows:
+On containerd, the windows platform matcher is responsible for pulling an appropriate manifest from an image index on windows platforms. Today, this platform matcher on windows always looks for a manifest with the OSVersion that matches the host and pulls it. This can be changed to pick a manifest with OSVersion that matches the one set for a particular runtime class in containerd.toml instead of the host OSVersion.
+An example of how the guest platform information can be set in containerd is as follows:
+
+<pre>
+        [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process]
+          base_runtime_spec = ""
+          cni_conf_dir = ""
+          cni_max_conf_num = 0
+          container_annotations = ["io.microsoft.container.*"]
+          pod_annotations = []
+          privileged_without_host_devices = false
+          privileged_without_host_devices_all_devices_allowed = false
+          runtime_path = ""
+          runtime_type = "io.containerd.runhcs.v1"
+          sandbox_mode = ""
+          snapshotter = ""
+
+<mark>
+          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process.guest_platform]
+            Architecture = "amd64"
+            OS = "windows"
+            OSFeatures = []
+            OSVersion = "10.0.20348"
+            Variant = ""
+</mark>
+          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process.options]
+</pre>
+
+On k8s, the pod and runtime class yamls would look as follows:
+
+<pre>
+apiVersion: node.k8s.io/v1
```

Comment:
Yes we would need a runtimeclass is OSVersion is different. Do you mean we should pass it in like this?
apiVersion: v1 
kind: Pod 
metadata: 
  name: mypod 
spec:
  containers: 
  - name: busybox 
    image: docker.io/library/busybox:1.35.0-musl 
    runtimeClassName: runhcs-wcow-runhcs

Reply:
@mrunalp could you explain a bit more what you mean? Setting the `runtimeClassName` on the pod spec is how this works today. 

There is some documentation on the various uses cases here https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1301-windows-runtime-class#story-2---forward-compatibility-with-hyper-v

Reply:
In the example above, OSVersion is `10.0.20348`. If there are multiple different OSVersions that need to be supported, mutliple RuntimeClasses will have to be created.

Reply:
We want this to be more generic for all platforms. For windows, this would be OSVersion but linux does not use OSVersion and if they need to override default behavior, they could override the variant information. The guestPlatform field mentioned in the example for containerd is of type ocispec.Platform . Passing osversion alone will not be generic enough

Reply:
container runtimes will use runtimeclasses to read the associated config from the toml and set up platform matchers (example in containerd) as needed.

Reply:
Thanks!

Reply:
Either way works for me... tuple or triplet  (image reference/digest, runtime class, platform (the oci platform struct)).  We can have more runtime classes one per unique desired platform (and name them accordingly), or pass the portions of the platform struct that the user desires to switch on, from the image spec portion(s) of the pod spec... Do all of the containers need to be in the same vm/type?

Reply:
@mikebrow "Do all of the containers need to be in the same vm/type?" not sure I understand this. All the containers inside of a pod have to be using the same runtime class. If a pod is started as hyperV isolated all the containers inside the pod would be hyperV isolated as well.

I think I would prefer having runtime class to pass this information for reasons mentioned by James.